segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -123,12 +130,34 @@
      * {@link ContainerEventProcessor.EventProcessor} based on its name.
      */
     @VisibleForTesting
-    static Function<String, CompletableFuture<DirectSegmentAccess>> getOrCreateInternalSegment(SegmentContainer container,
-                                                                                               MetadataStore metadataStore,
-                                                                                               Duration timeout) {
-        return s -> metadataStore.registerPinnedSegment(getEventProcessorSegmentName(container.getId(), s),
-                SYSTEM_CRITICAL_SEGMENT, null, timeout) // Segment should be pinned.
-                .thenCompose(l -> container.forSegment(getEventProcessorSegmentName(container.getId(), s), timeout));
+    Function<String, CompletableFuture<DirectSegmentAccess>> getOrCreateInternalSegment(SegmentContainer container,
+                                                                                        MetadataStore metadataStore,
+                                                                                        Duration timeout) {
+        return s -> tryCreateInternalSegment(container, s, timeout)
+                .thenCompose(v -> container.forSegment(getEventProcessorSegmentName(container.getId(), s), timeout))
+                .thenApply(directAccessSegment -> {
+                    metadataStore.pinSegmentToMemory(getEventProcessorSegmentName(container.getId(), s), timeout); // Segment should be pinned.
+                    return directAccessSegment;
+                });
+    }
+
+    private CompletableFuture<Void> tryCreateInternalSegment(SegmentContainer container, String eventProcessorName, Duration timeout) {
+        synchronized (eventProcessorMap) {
```

Comment:
Does this need to be synchronized?  It will end up serializing on each and every call to `ContainerEvenProcessor`.
`eventProcessorMap `is already `ConcurrentHashMap `, so may be `computeIfAbsent `is better?

Reply:
Maybe not, as we are just checking the content of that entry and not modifying anything. I have removed the synchronized block and changed the condition to do just one get.

Reply:
ok I missed  that call is made from synchronised block in createEventProcessor